File: ts/solutions/linked/linked_7.ts
import { LinkedList, Node_ } from './linked_1'

// see Remove Duplicates for List and Node type definitions

// O(m + n)
function findIntersection(l1: LinkedList, l2: LinkedList): Node_ {
  const nodeSet: Set<Node_> = new Set()

  let node = l1.head
  while (node) {
    nodeSet.add(node)
    node = node.next
  }

  node = l2.head
  while (node) {
    if (nodeSet.has(node)) {
      return node
    }
    node = node.next
  }

  return null
}

File: go/pkg/linked/linked_7.go
package linked

// FindIntersection is O(m + n)
func FindIntersection(l1, l2 List) *Node {
	m := map[*Node]bool{}

	node := l1.Head
	for node != nil {
		m[node] = true
		node = node.Next
	}

	node = l2.Head
	for node != nil {
		if m[node] {
			return node
		}
		node = node.Next
	}

	return nil
}
